So, first thing first. Ill start examining a radom point of New York, and see how dense is that area with a query to Foursquare API. The expected result is that New York its a totally hot area terribly expensive, and as such, not a safe bet for someone who is taking its first steps in the international scene. High risk high reward surely, but it has no meaning to throw yourself to the sharks like that, and ill try to prove it through data:

In [ ]:
#Import necesary libraries for the sake of this particular study:



In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


In [2]:
#Ill save the keys necesary to call the Foursquare API later, as shown in the course:
CLIENT_ID = '0ETFVTQEUWSRRP2EQ5JHZW0UUC4AMRXYX0YFFHQAKZLJJD10' # your Foursquare ID
CLIENT_SECRET = 'T4K2MZTXLDEVIPIL15LTDI3QPNBSHMYAIHXBCLNBSFFEH1FH' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 52
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 0ETFVTQEUWSRRP2EQ5JHZW0UUC4AMRXYX0YFFHQAKZLJJD10
CLIENT_SECRET:T4K2MZTXLDEVIPIL15LTDI3QPNBSHMYAIHXBCLNBSFFEH1FH


In [3]:
#As a random point in a relevant area, i will reutilize the geolocator to the exact area as shown in the course:
address = '102 North End Ave, New York, NY'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address, timeout=10)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

40.7149555 -74.0153365


In [89]:
#Ill start the query using a ratio of 2500, expecting the result to surpass the 49 unfortunately limit of the API, using as keyword 'pizza':
search_query = 'pizza'
radius = 10000
print('done!')

done!


In [190]:
#Call the API:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=0ETFVTQEUWSRRP2EQ5JHZW0UUC4AMRXYX0YFFHQAKZLJJD10&client_secret=T4K2MZTXLDEVIPIL15LTDI3QPNBSHMYAIHXBCLNBSFFEH1FH&ll=40.36,-74.01&v=20180604&query=bank&radius=5000&limit=52'

In [197]:
#Request the information and display it properly:
results = requests.get(url).json()

In [192]:
#Normalize the info as shown over the course and transform it to a dataframe, displaying the first 3 results:
venues = results['response']['venues']
dataframe = json_normalize(venues)
dataframe.head(3)

categories delivery.id  \
0  [{'id': '4bf58dd8d48988d13d941735', 'name': 'H...         NaN   
1  [{'id': '52f2ab2ebcbc57f1066b8b46', 'name': 'S...         NaN   
2  [{'id': '4bf58dd8d48988d129951735', 'name': 'T...         NaN   

  delivery.provider.icon.name delivery.provider.icon.prefix  \
0                         NaN                           NaN   
1                         NaN                           NaN   
2                         NaN                           NaN   

  delivery.provider.icon.sizes delivery.provider.name delivery.url  hasPerk  \
0                          NaN                    NaN          NaN    False   
1                          NaN                    NaN          NaN    False   
2                          NaN                    NaN          NaN    False   

                         id  location.address     ...       \
0  4be16a84d816c928d5f7efd9      101 Ridge Rd     ...        
1  4b9a41e4f964a520c6a735e3      362 Broad St     ...        
2  4af38fc5f964a5209fee21e3  175 Monmouth St.     ...        

                           location.formattedAddress  \
0  [101 Ridge Rd, Little Silver, NJ 07739, United...   
1  [362 Broad St, Red Bank, NJ 07701, United States]   
2  [175 Monmouth St., Red Bank, NJ 07701, United ...   

                             location.labeledLatLngs location.lat  \
0  [{'label': 'display', 'lat': 40.35016004059454...    40.350160   
1  [{'label': 'display', 'lat': 40.339588, 'lng':...    40.339588   
2  [{'label': 'display', 'lat': 40.34828085719662...    40.348281   

  location.lng  location.neighborhood location.postalCode location.state  \
0   -74.039729                    NaN               07739             NJ   
1   -74.064127                    NaN               07701             NJ   
2   -74.074357                    NaN               07701             NJ   

                            name    referralId venuePage.id  
0  Red Bank Regional High School  v-1549514468          NaN  
1     Super Foodtown of Red Bank  v-1549514468    209529457  
2  NJT - Red Bank Station (NJCL)  v-1549514468          NaN  

[3 rows x 25 columns]

In [93]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name           categories  \
0             Umberto Restaurant & Pizza          Pizza Place   
1          Urban CoalHouse Pizza and Bar          Pizza Place   
2                MJ's Pizza, Bar & Grill  American Restaurant   
3    Luigi's Famous Pizza and Ristorante          Pizza Place   
4                            Earth Pizza          Pizza Place   
5                 MJ's Pizza Bar & Grill  American Restaurant   
6             Patrizia's Pizza and Pasta   Italian Restaurant   
7   Mangia Brick Oven Pizza - Shrewsbury          Pizza Place   
8                        Mr. Pizza Slice          Pizza Place   
9                Zoni's Coal Fired Pizza          Pizza Place   
10          Attilio's Restaurant & Pizza          Pizza Place   
11                  Luigi's Famous Pizza          Pizza Place   
12          Manny's Pizza, Subs & Salads          Pizza Place   
13       Joe's Pizza And Cucina Italiana          Pizza Place   
14              Federico's Pizza Express          Pizza Place   
15      Merendino's Pizza and Restaurant          Pizza Place   
16         Maurizio's Pizza & Ristorante          Pizza Place   
17                       Basille's Pizza          Pizza Place   
18                         Pizza Company          Pizza Place   
19                        1st Ave. Pizza          Pizza Place   
20           Gigi's Pizza And Restaurant          Pizza Place   
21               Brother's Pizza & Pasta          Pizza Place   
22                         Bella's Pizza          Pizza Place   
23                  Luigi's Famous Pizza          Pizza Place   
24                     Chiafullo's Pizza          Pizza Place   
25                             Pizza Mia   Italian Restaurant   
26                      New Corner Pizza   Italian Restaurant   
27           Gigi's Pizza And Restaurant          Pizza Place   
28               Abramos Pizza And Pasta          Pizza Place   
29                       Palumbo's Pizza          Pizza Place   
30    Olivia's Trattoria & Artisan Pizza           Restaurant   
31                             Pizza MIA                 None   
32              Tommy's Coal Fired Pizza          Pizza Place   
33                         Pizza Company          Pizza Place   
34                        Domino's Pizza          Pizza Place   
35                        Domino's Pizza          Pizza Place   
36                 Pizza & Pasta Factory          Pizza Place   
37                         Danny's Pizza          Pizza Place   
38               Zoni's Coal Fired Pizza           Restaurant   
39                     new corners pizza   Italian Restaurant   
40                          Scenic Pizza          Pizza Place   
41       Attilios Pizza West Long Branch   Italian Restaurant   
42                      Fratelli's Pizza          Pizza Place   
43  Maria's Pizza And Italian Restaurant           Restaurant   
44                      Federico's Pizza          Pizza Place   
45                     Francesco's pizza          Pizza Place   
46               Zoni's Coal Fired Pizza          Pizza Place   
47                        Broadway Pizza          Pizza Place   
48                        Red bank pizza                 None   
49                     Papa John's Pizza          Pizza Place   

                                  address  cc                city  \
0                            583 River Rd  US          Fair Haven   
1                            2 Bridge Ave  US            Red Bank   
2                     1005 State Route 35  US          Middletown   
3                        86 Oceanport Ave  US       Little Silver   
4                             95 Broad St  US            Red Bank   
5                       1213 Sycamore Ave  US        Tinton Falls   
6                             28 Broad St  US            Red Bank   
7                           1087 Broad St  US          Shrewsbury   
8                          10 Monmouth St  US            Red Bank   
9                            2

As expected, the query surpass the 49 limits of the API, not good. Lets visualize the results, and later decrease some the radio to display more consistent results:

In [11]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[latitude, longitude],
                        zoom_start=13.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

Surely its not for sale!
Lets decrease the radio and display the results again...

In [12]:
search_query = 'pizza'
radius = 1000
print('done!')

done!


In [17]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[latitude, longitude],
                        zoom_start=13.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

49 results again, showing the overpopulation of places related to pizza, wont even imagine the scenario asking a more generic 'restaurants'

So, lets move around a little. Lets see in another corner of the country, say 7th avenue with 86th street, pretty close with lot of movement also. 

In [52]:
search_query = 'pizza'
radius = 5000
print('done!')

done!


In [23]:
latitude=40.62
longitude=-74.03
print("ready to go!")

ready to go!


In [28]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name  \
0                                  Espresso Pizza   
1                                 Verrazano Pizza   
2                                     Pizza Wagon   
3              Zitelli's Thin Crust Pizza and Bar   
4                                   Luigi's Pizza   
5                       Artichoke Basille's Pizza   
6                            Papa's Pizza & Pasta   
7                         Krispy Pizza - Brooklyn   
8                                   Bklyn's Pizza   
9                                    Nino's Pizza   
10                             Original Pizza III   
11     Halal Food Stand (across from Pizza Wagon)   
12                      Campania Coal Fired Pizza   
13                                 Domino's Pizza   
14                                   Bennys Pizza   
15                      artichoke basille's pizza   
16                                  Nonno's Pizza   
17                           Vaccaro Pizza Royale   
18                     Mancini's Wood-Fired Pizza   
19                       Pizza Supremo Restaurant   
20                                Bay Ridge Pizza   
21                                    Pizza House   
22                               Buon Gusto Pizza   
23         Giacomo's Wood Fired Pizza & Trattoria   
24                                   Decora Pizza   
25                                  L and J Pizza   
26                            J&A Pizza and pasta   
27                                  Barri's Pizza   
28  Jean Danet  Pastry, Cafe and Brick Oven Pizza   
29                                   Bari's Pizza   
30                                      Pizza Den   
31                              Mias Pizza Shoppe   
32                                     Pizza City   
33                                  Landy's Pizza   
34                                Benevento Pizza   
35                                    Pizza Wagon   
36                             Original Pizza 111   
37                           michael & sons pizza   
38                        Molto Bene Cheese Pizza   
39                                   Aurora Pizza   
40                               Brick Oven Pizza   
41                                  Mario's Pizza   
42                     Angie's Pizza & Restaurant   
43                                  Tom's Pizza 3   
44                       California Pizza Kitchen   
45                    New York Pizza and Macaroni   
46                              Pan Pizza Kitchen   
47                            Philadelphia Grille   
48                                  Tom's Pizza 2   
49                              Papa John's Pizza   

                     categories                  address  cc           city  \
0                   Pizza Place             9403 5th Ave  US       Brooklyn   
1                   Pizza Place             9102 4th Ave  US       Brooklyn   
2                   Pizza Place             8610 5th Ave  US       Brooklyn   
3            Italian Restaurant          8530-3rd Avenue  US       Brooklyn   
4                   Pizza Place             8502 3rd Ave  US       Brooklyn   
5                   Pizza Place                      NaN  US       Brooklyn   
6                   Pizza Place             8312 5th Ave  US       Brooklyn   
7                   Pizza Place            7112 13th Ave  US       Brooklyn   
8                   Pizza Place             1406 86th St  US       Brooklyn   
9                   Pizza Place             9110 3rd Ave  US       Brooklyn   
10                  Pizza Place             6823 4th Ave  US       Brooklyn   
11    Middle Eastern Restaurant                      NaN  US       Brooklyn   
12           Italian Restaurant             9824 4th Ave  US       Brooklyn   
13                  Pizza Place             9105 4th Ave  US       Brooklyn   
14                  Pizza Place                      NaN  US       Brooklyn   
15      New American Restaurant             9102 4th Ave  US       Brooklyn  

I wont put all the steps again cause it is a pretty iterative process, and this notebook would get incredibly huge. But above is the result for the new query, 49 results again... lets see them.

In [29]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[latitude, longitude],
                        zoom_start=13.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

Ok, totally "overpopulated" area. Lets move around a little. Unfortunately, this process would be much more efficient if the limit of results from Foursquare was not as it is now, but les go ahead. Lets move more to the south, working a little bit with google earth. 

In [30]:
latitude=40.45
longitude=-74.22
print("ready to go!")

ready to go!


In [41]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name          categories  \
0                          V & F Pizza Restaurant         Pizza Place   
1                  Villa Bella Pizza & Restaurant         Pizza Place   
2                       Tony's Pizza & Restaurant         Pizza Place   
3                      Keyport Pizza & Restaurant         Pizza Place   
4                                Ciros Pizza Cafe         Pizza Place   
5                            Denino's Pizza Place         Pizza Place   
6                             Sal And Ginos Pizza         Pizza Place   
7                                Ray's Real Pizza  Italian Restaurant   
8                                Giuseppe's Pizza         Pizza Place   
9                                   Pizza by kims                None   
10                               Esposito's Pizza         Pizza Place   
11                        Tony's Pizza Restaurant         Pizza Place   
12                          Carlo's Gourmet Pizza         Pizza Place   
13                             Crazy Mike's Pizza         Pizza Place   
14                    Flame800 - Coal Fired Pizza         Pizza Place   
15                                 Angelo's Pizza         Pizza Place   
16                                          Pizza         Pizza Place   
17                                 Domino's Pizza         Pizza Place   
18                                 Domino's Pizza         Pizza Place   
19                                 Angela's Pizza         Pizza Place   
20                           Dylan's Pizza Parlor  Miscellaneous Shop   
21                                La Strada Pizza         Pizza Place   
22                             Gourmet Pizza City         Pizza Place   
23                             Gourmet Pizza City         Pizza Place   
24                           Denino's Pizza Place         Pizza Place   
25                                   Pizza Planet         Pizza Place   
26               Lockport Pies Pizza & Restaurant         Pizza Place   
27                        Pizza and pasta factory                None   
28                                    Pizza Loolu         Pizza Place   
29                              Papa John's Pizza         Pizza Place   
30                    Romeo's pizza pasta express         Pizza Place   
31                                   Scotto Pizza         Pizza Place   
32                               Baris Pizza Cafe         Coffee Shop   
33                                 Domino's Pizza         Pizza Place   
34                    Brunswick Zone Hazlet Lanes       Bowling Alley   
35                Nicco's Pizza & Italian Cuccina         Pizza Place   
36                                  attilio pizza         Pizza Place   
37                                  Romeo's Pizza         Pizza Place   
38  A. Scotto's Pizzaeria & Ristorante Napoletana         Pizza Place   
39                     Italian Dans Pasta & Pizza  Italian Restaurant   
40                              Union Beach Pizza        Neighborhood   
41                              Papa John's Pizza         Pizza Place   
42                               Mondello's Pizza  Italian Restaurant   
43             Lisa Italian Restaurant & Pizzeria  Italian Restaurant   
44                                    Joe's Pizza         Pizza Place   
45                                 Todaro's Pizza  Italian Restaurant   
46                    il fornetto coal fire pizza         Pizza Place   
47                                  Dominos Pizza         Pizza Place   
48                              Papa John's Pizza         Pizza Place   
49                            Vivaldis Pizza Cafe         Pizza Place   

                address  cc             city        country       crossStreet  \
0   309A State Route 35  US        Cliffwood  United States               NaN   
1                   NaN  US        Cliffwood  United States               NaN   
2    300 State Route 35  US          Keyport  United States               NaN 

In [43]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.45, -74.22],
                        zoom_start=11.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

Ok, so 49 results again. A hot area again. Lets move again, to see if there is a less congested area 

In [58]:
latitude=40.31
longitude=-73.98
print("ready to go!")

ready to go!


In [64]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.31, -73.98],
                        zoom_start=11.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

Wrong call again. Dont worry, this land somewhere. If one looks closely, the distribution is really dense there. Much less dispersion that in the other cases. So it seems that moving up to say Highlands or similar, can finally give a promising result. 

In [65]:
latitude=40.397
longitude=-73.986
print("ready to go!")

ready to go!


In [71]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.397, -73.986],
                        zoom_start=12.4,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

So finally something. For a radio of 5000, just 13 results. We land in a Borough called Highlands, that seems promising. It is near the see, it doesnt seem to be on the desert, and its between really hot spots such as Long Branch as seen before, and Keyport -thee map that is just before Long Branch. It actually seems like Highlands have some movement after all, but Fair Haven, for example, has really not movements. Lets do the final map to see if moving a little bit toward Fair Heaven, reveals more information. Also, it would be nice to get the whole picture of the area. So lets zoom out a little more, and then we can proceed to move toward exploring the place founded.

In [88]:
latitude=40.36
longitude=-74.01
print("ready to go!")

ready to go!


In [96]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.36, -74.01],
                        zoom_start=12,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

So, it seems that there are other hot spots, with high density, but Rumson Road remains with no movement at all. Finally, it seems, that this place is totally an option. At least if its not a ghost town or something. It seem that the untoched area is divided between Fair Haven and Rumson. Lets do some research about them.

In [97]:
from lxml import html
import requests

In [98]:
page = requests.get('https://en.wikipedia.org/wiki/Rumson,_New_Jersey')

In [99]:
tree=html.fromstring(page.content)

In [100]:
from bs4 import BeautifulSoup

In [102]:
soup = BeautifulSoup(page.text, 'html.parser')

In [141]:
import requests
from bs4 import BeautifulSoup
 
def count_words(url, the_word):
    r = requests.get(url, allow_redirects=False)
    soup = BeautifulSoup(r.content, 'lxml')
    words = soup.find(text=lambda text: text and the_word in text)
    print(words)
    return len(words)
 
def main():
    url = 'https://en.wikipedia.org/wiki/Rumson,_New_Jersey'
    word = 'income'
    count = count_words(url, word)
    print('\nUrl: {}\ncontains {} occurrences of word: {}'.format(url, count, word))
 
if __name__ == '__main__':
    main()

median household income

Url: https://en.wikipedia.org/wiki/Rumson,_New_Jersey
contains 23 occurrences of word: income


In [ ]:
So, les find manually the relevant words: median household income

In [ ]:
print(soup.prettify())

In [144]:
The median household income was of 134,281 which seems pretty high. Good news. 

In [149]:
import requests
from bs4 import BeautifulSoup
 
def main():
    url = 'https://en.wikipedia.org/wiki/Rumson,_New_Jersey'
    word = 'population'
    count = count_words(url, word)
    
if __name__ == '__main__':
    main()

, the borough's population was 7,122,


So there you go, a population of 7,122 for the community, not much but still close to other locations quite populated. So, lets dig a bit more. But first, we should look a bit more from the county where Rumson belongs, to see, at least, if it is not worst around...  

In [166]:
import requests
from bs4 import BeautifulSoup
 
def main():
    url = 'https://en.wikipedia.org/wiki/Monmouth_County,_New_Jersey'
    word = 'population was'
    count = count_words(url, word)
    
if __name__ == '__main__':
    main()

, the county's population was 626,351, making it the state's fifth-most populous county,


In [170]:
import requests
from bs4 import BeautifulSoup
 
def main():
    url = 'https://en.wikipedia.org/wiki/Monmouth_County,_New_Jersey'
    word = 'income'
    count = count_words(url, word)
    
if __name__ == '__main__':
    main()

per capita personal income


So, one must iterate the process to search for the key word, and see if there is any luck in the process. This systems does not always work, but is worth the try

In [171]:
page = requests.get('https://en.wikipedia.org/wiki/Monmouth_County,_New_Jersey')

In [172]:
soup = BeautifulSoup(page.text, 'html.parser')

In [ ]:
print(soup.prettify())

<a href="/wiki/Per_capita_personal_income_in_the_United_States" title="Per capita personal income in the United States">
        per capita personal income
       </a>
       of $69,410, the fifth-highest in New Jersey and ranked 74th of 3,113 counties in USA
       

So, not bad at all. The county its quite wealthy and Rumson income is even higher ($79,388 per capita). So, it is located in quite a financially healthy place. 

The next step its to work a little on the presentation to be shown to the italian friend. We seem to have a little of luck this time. But lets dig closer with more queries to the relevant spot, to understand more about its infraestructure.

In [177]:
#Ill start the query using a ratio of 2500, expecting the result to surpass the 49 unfortunately limit of the API, using as keyword 'pizza':
search_query = 'hospital'
radius = 5000
print('done!')

done!


In [181]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name      categories          address  cc  \
0     Fair Haven Animal Hospital    Veterinarian              NaN  US   
1  Hospital For Wayward Children            None    15 Lisa Court  US   
2      Monmouth Medical Hospital  Shop & Service              NaN  US   
3       Riverview Medical Center        Hospital  1 Riverview Plz  US   
4  Little Silver Animal Hospital    Veterinarian              NaN  US   
5                 Two Rivers Vet    Veterinarian     457 River Rd  US   
6       Hospitalty Room #2 - MMC        Hospital              NaN  US   
7    Riverview Hospital Dialysis  Medical Center    48 E Front St  US   

            city        country  distance  \
0     Fair Haven  United States      2072   
1            NaN  United States      4762   
2            NaN  United States      3171   
3       Red Bank  United States      4635   
4  Little Silver  United States      4491   
5     Fair Haven  United States      3184   
6    Long Branch  United States      6254   
7       Red Bank  United States      4596   

                                    formattedAddress  \
0                    [Fair Haven, NJ, United States]   
1         [15 Lisa Court, New Jersey, United States]   
2                        [New Jersey, United States]   
3  [1 Riverview Plz, Red Bank, NJ 07701, United S...   
4           [Little Silver, NJ 07739, United States]   
5  [457 River Rd, Fair Haven, NJ 07704, United St...   
6             [Long Branch, NJ 07740, United States]   
7  [48 E Front St, Red Bank, NJ 07701, United Sta...   

                                      labeledLatLngs        lat        lng  \
0  [{'label': 'display', 'lat': 40.365799, 'lng':...  40.365799 -74.033221   
1  [{'label': 'display', 'lat': 40.337375, 'lng':...  40.337375 -74.057638   
2  [{'label': 'display', 'lat': 40.35266716169504...  40.352667 -73.973869   
3  [{'label': 'display', 'lat': 40.35351484308679...  40.353515 -74.063975   
4  [{'label': 'display', 'lat': 40.32846459215351...  40.328465 -74.043031   
5  [{'label': 'display', 'lat': 40.35919951666666...  40.359200 -74.047529   
6  [{'label': 'display', 'lat': 40.30462914823733...  40.304629 -73.997485   
7  [{'label': 'display', 'lat': 40.35266705430916...  40.352667 -74.063325   

  postalCode       state                        id  
0        NaN          NJ  514b3927e4b05ed6ecbc417e  
1        NaN  New Jersey  4d143aedbb488cfa0c6695d4  
2        NaN  New Jersey  4e57ce9da8096558610db2e7  
3      07701          NJ  4b608208f964a5207bea29e3  
4      07739          NJ  4d0b958b9a443704c09cba67  
5      07704          NJ  4cbb461ff50e224b81bbf8fb  
6      07740          NJ  4f0121b86da1cd035a0e9132  
7      07701          NJ  4c6bee1b96eaa593007123df

In [182]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.36, -74.01],
                        zoom_start=12,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

In [183]:
#Ill start the query using a ratio of 2500, expecting the result to surpass the 49 unfortunately limit of the API, using as keyword 'pizza':
search_query = 'school'
radius = 5000
print('done!')

done!


In [187]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name                    categories  \
0                The Rumson Country Day School                Private School   
1                Rumson Fair Haven High School                   High School   
2                           Rumson High School                          None   
3                    School of Rock - Red Bank                  Music School   
4                           Forrestdale School                        School   
5     Rumson Fair haven high school auditorium                    Auditorium   
6                Red Bank Regional High School                   High School   
7                             Knollwood School     College Academic Building   
8                    Sickles Elementary School             Elementary School   
9                 Point Road Elementary School             Elementary School   
10                        Meadow Flower School                        School   
11                                 School Park                    Playground   
12                      Red Bank Middle School                 Middle School   
13                        Markham Place School                        School   
14                    St. Johns Nursery School                       Daycare   
15                   Maple Place Middle School                 Middle School   
16                       Monmouth Beach School  General College & University   
17                           Holy Cross School                        School   
18                   Navisink Elementry School             Elementary School   
19           Voting Booth Markham Place School                  Voting Booth   
20           SBU: Long Branch Satellite School        College Residence Hall   
21                           Tower Hill School                        School   
22               Red Bank Catholic High School                   High School   
23                     Red bank charter school                          None   
24                             Navesink School                        School   
25                     Red Bank Driving School                Driving School   
26                         The Language School             College Classroom   
27                       Audrey W Clark School             Elementary School   
28                 Wolf Hill Elementary School             Elementary School   
29  A Step in the Right Direction Dance School                  Dance Studio   
30      Amerigo A. Anastasia Elementary School             Elementary School   
31                             Seashore School                  College Quad   
32           "Kids Who Rock", Rock Band School         Performing Arts Venue   
33     Tinton Falls Cooperative Nursery School                Nursery School   
34                   Gregory School Playground                    Playground   
35                            Anastasia School                        School   
36              Champions at Conrow Elementary             Elementary School   
37                 Highlands Elementary School             Elementary School   
38                 St. James Elementary School             Elementary School   
39                   Gregory Elementary School             Elementary School   

                       address  cc                city        country  \
0                          NaN  US              Rumson  United States   
1                          NaN  US              Rumson  United States   
2                          NaN  US              Rumson  United States   
3               52 Monmouth St  US            Red Bank  United States   
4                  Forrest Ave  US              Rumson  United States   
5                          NaN  US              Rumson  United States   
6                 101 Ridge Rd  US       Little Silver  United States   
7                 Hance Street  US          Fair Haven  United States   
8                          NaN  US          Fair Haven  United States   
9   357 Little Silver Point Rd 

In [188]:
import pandas as pd
import folium
from folium.plugins import FastMarkerCluster, MarkerCluster

df = dataframe_filtered

folium_map = folium.Map(location=[40.36, -74.01],
                        zoom_start=12,
                        tiles='Stamen Terrain')

mc = MarkerCluster(name="Marker Cluster")

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(folium_map)

mc.add_to(folium_map)

folium.LayerControl().add_to(folium_map)
folium_map

In [189]:
#Ill start the query using a ratio of 2500, expecting the result to surpass the 49 unfortunately limit of the API, using as keyword 'pizza':
search_query = 'bank'
radius = 5000
print('done!')

done!


So, lots of schools and hospital arounds. As seen before, sourrunded by very hot spots, and at the same time, with lots of potencial consumers with quite a wealthy situation. Seems like a safe bet. 